src/theory/term_registration_visitor.cpp

```diff
@@ -47,62 +87,16 @@ bool PreRegisterVisitor::alreadyVisited(TNode current, TNode parent) {
     Debug("register::internal") << "quantifier:true" << std::endl;
     return true;
   }
-
-  TheoryId currentTheoryId = Theory::theoryOf(current);
-  TheoryId parentTheoryId = Theory::theoryOf(parent);
-
-  d_theories = TheoryIdSetUtil::setInsert(currentTheoryId, d_theories);
-  d_theories = TheoryIdSetUtil::setInsert(parentTheoryId, d_theories);
-
-  // Should we use the theory of the type
-  bool useType = false;
-  TheoryId typeTheoryId = THEORY_LAST;
-
-  if (current != parent) {
-    if (currentTheoryId != parentTheoryId) {
-      // If enclosed by different theories it's shared -- in read(a, f(a)) f(a) should be shared with integers
-      TypeNode type = current.getType();
-      useType = true;
-      typeTheoryId = Theory::theoryOf(type);
-    } else {
-      TypeNode type = current.getType();
-      typeTheoryId = Theory::theoryOf(type);
-      if (typeTheoryId != currentTheoryId) {
-        if (type.isInterpretedFinite()) {
-          useType = true;
-        }
-      }
-    }
-  }
   
   // Get the theories that have already visited this node
   TNodeToTheorySetMap::iterator find = d_visited.find(current);
   if (find == d_visited.end()) {
-    if (useType) {
-      d_theories = TheoryIdSetUtil::setInsert(typeTheoryId, d_theories);
-    }
+    // not visited at all, return false
     return false;
   }
 
   TheoryIdSet visitedTheories = (*find).second;
-  if (TheoryIdSetUtil::setContains(currentTheoryId, visitedTheories))
-  {
-    // The current theory has already visited it, so now it depends on the parent and the type
-    if (TheoryIdSetUtil::setContains(parentTheoryId, visitedTheories))
-    {
-      if (useType) {
-        TheoryId typeTheoryId2 = Theory::theoryOf(current.getType());
-        d_theories = TheoryIdSetUtil::setInsert(typeTheoryId2, d_theories);
-        return TheoryIdSetUtil::setContains(typeTheoryId2, visitedTheories);
-      } else {
-        return true;
-      }
-    } else {
-      return false;
-    }
-  } else {
-    return false;
-  }
+  return !shouldVisit(visitedTheories, current, parent);
```

Comment:
It is a bit confusing to me that "alreadyVisited" depends on "shouldVisit". I would expect it to be the other way around. Can you clarify this in a commend, namely -- explain the call the `shouldVisit`?

Reply:
I renamed the method to `isAlreadyVisited`.